### DISCUSSION TITLE
When to dispose async engine?

### DISCUSSION THREAD

**Discussion by jocho-here** [sqlalchemy/sqlalchemy#7531] [Usage Questions] [answered]
Hi,

my API server creates async engine and assign that to a global variable.

Whenever there's a new request, it then runs `async with engine.begin() as conn` and passes the connection of the engine when running queries.

I'm reading the documentation and it says that I should dispose the engine at the end of the function if I've created the engine for the function scope.

My questions are...

1) Is it okay for multiple requests to share a single `AsyncEngine`? They will be taken care of by a single thread.
2) When should I dispose the `AsyncEngine`? Just looking at how my engine creation is set up, it looks like it's proper to call `engine.dispose()` right before the API server dies.

Any suggestions?

**Comment by CaselIT** [comment, marked as answer]
Hi,

Usually there is a single engine in an application, this engine can be shared between multiple threads.
You can dispose the engine before the application exits

**Reply by jocho-here** [reply]
Sounds good. Thanks!

**Reply by filippaquant** [reply]
Hey, thanks for an answer.
But I have an extra question.
How can I dispose the engine that is not exist? I mean at the creation of the instance I don't have an engine to dispose

**Reply by CaselIT** [reply]
sorry, but I'm not sure I understand your question. Can you provide an example?

**Reply by filippaquant** [reply]
I mean to run `dispose()` on an object I need to have it created, right? 
Maybe I don't understand something.
Here is a small example:

```
class AsyncEngine:
    
    def __init__(self, uri, async_uri):
        self.async_engine = create_async_engine(async_uri)
        self.engine = create_engine(uri)

    def get(self):
        with self.engine.begin() as conn:
            pass

    async def aget(self):
        async with self.async_engine.begin() as conn:
            pass

```

```
if __name__ == '__main__':
    # Don't have a link to async_engine here to dispose
    async_engine = AsyncEngine("URI", "ASYNC_URI")`
```

Before I create a new AsyncEngine, I don't have an access to an engine field itself and can't `dispose()` it because it's not yet created.
And afterwords sync engine will be garbage collected, but `async_engine` won't. According to sqlalchemy docs:

> It’s advisable to invoke the [AsyncEngine.dispose()](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncEngine.dispose) method using await when using the [AsyncEngine](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncEngine) object in a scope that will go out of context and be garbage collected, as illustrated in the async_main function in the above example. This ensures that any connections held open by the connection pool will be properly disposed within an awaitable context. Unlike when using blocking IO, SQLAlchemy cannot properly dispose of these connections within methods like __del__ or weakref finalizers as there is no opportunity to invoke await. Failing to explicitly dispose of the engine when it falls out of scope may result in warnings emitted to standard out resembling the form RuntimeError: Event loop is closed within garbage collection.

[https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#synopsis-core](url)

**Reply by CaselIT** [reply]
usually the async application are either
- web server, where there is an [event](https://asgi.readthedocs.io/en/latest/specs/lifespan.html#lifespan-protocol) that's sent when the server is exiting (this event may be exposed with some high level api depending on the asgi framework in use)
- are cli application that run using `asyncio.run(my_coro)`. In that case before `my_coro` returns add a call to dispose the async engine

in either case you should have access to the async engine so you can call dispose
